### ISSUE TITLE
'when' statement should be evaluated before anything else to avoid errors in other statements

### ISSUE THREAD

**Issue by maxalbert** [ansible/ansible#10400]
**Issue Type:** Bug Report

**Ansible Version:** 1.8.4

**Environment:** N/A

**Summary:**

The `when` statement should be evaluated before anything else and the task be skipped immediately if it evaluates to False. Otherwise it may cause failures in other statements (e.g. `with_sequence` in the example below).

**Steps To Reproduce:**

Here is a minimal playbook `bug.yml` to illustrate the issue:

``` yaml

---
- hosts: all
  gather_facts: False
  tasks:
    - name: "This task should be skipped but it fails instead."
      command: echo "Hello world"
      when: False
      with_sequence: start=2 end=1
```

It can be run with:

`ansible-playbook -c local -i localhost, bug.yml`

**Expected Results:**

The task should be skipped because the `when` statement evaluates to False. The expected output is:

```
TASK: [This task should be skipped but it fails instead.] *** 
skipping: [localhost]
```

**Actual Results:**

The task fails because the start value in the `with_sequence` statement is larger than the end value.

```
TASK: [This task should be skipped but it fails instead.] *** 
fatal: [localhost] => can't count backwards

FATAL: all hosts have already failed -- aborting
```

**Comment by srgvg** [comment]
I disagree this is a bug.

`when:` conditionals are evaluated per-item in the loop. This has always been the expected behaviour.
I, for example, have lots of tasks where I use this explicitly.

I can however understand there are use cases where you want a general condition. This might be an idea for a feature request.

**Comment by maxalbert** [comment]
Thanks for the quick reply! It didn't occur to me that the conditional is evaluated per-item. I just saw that this is indeed mentioned in the section on [Conditionals](http://docs.ansible.com/playbooks_conditionals.html) (even if it is somewhat buried), but it might be worth mentioning it in the section on [Loops](http://docs.ansible.com/playbooks_loops.html), too (which is where I looked). If I wanted to submit a PR for this, where in that section should the information go (e.g. right at the top or in some subsection)?

Regarding your suggestion for a feature request, what could this look like? Would it require introducing a completely new keyword, or is it possible to somehow modify the behaviour of the existing `when:` or loop-related statements to enable a general condition?

**Comment by srgvg** [comment]
I'm just a humble contributor. I'll leave it to the core team to answer your questions :)

**Comment by bcoca** [comment]
@maxalbert as per the documentation request, maybe a note under the standard loops example?

As for a 'preemptive when' I'm not sure how we would name that feature or how many people would need it. Closing this request as it is mostly a misunderstanding about the feature.

In the future, this might be a topic more well suited for the user list, which you can also post here if you'd like some more help with the above.
- https://groups.google.com/forum/#!forum/ansible-project
- https://groups.google.com/forum/#!forum/ansible-devel

Thank you once again for this and your interest in Ansible!

**Comment by maxalbert** [comment]
Ok, thanks again for the clarification and incredibly quick replies! I actually like the name `preemptive_when`. ;-) But I also don't know how generally useful it would be.

In my particular case I realised that the true shortcoming is that the `with_sequence` statement is not flexible enough. For example, it doesn't allow zero counts and cannot skip tasks if `start > end` (i.e. if the range is empty). I just opened separate issues for those: #10401, #10403. If this can be addressed then I'm happy without a `preemptive_when` statement.
